File: ios/Sources/SDUIRender/Theme.swift
#if canImport(SwiftUI)
import SwiftUI
import SDUICore

/// Resolves contract values (`Dimension`, hex colors, typography tokens) into
/// native SwiftUI types, using the shared token table carried by the binding
/// context. All the platform-specific mapping lives here so the renderers stay
/// declarative.
public enum Theme {

    // MARK: Dimensions

    public static func cgFloat(_ dim: SDUICore.Dimension?, ctx: BindingContext, default def: CGFloat = 0) -> CGFloat {
        guard let dim else { return def }
        switch dim {
        case .points(let p): return CGFloat(p)
        case .infinity:      return .infinity
        case .token(let ref):
            return CGFloat(BindingEngine.resolve(ref, in: ctx).doubleValue ?? Double(def))
        }
    }

    // MARK: Colors

    /// Accepts a token ref (`$token.color.primary`), a hex string, or a binding.
    /// When `$env.theme` is `dark`, a `$token.color.*` ref prefers its
    /// `$token.colorDark.*` counterpart if one exists — dark mode is a separate
    /// palette, not an inversion (see docs/design-notes.md).
    public static func color(_ raw: String?, ctx: BindingContext) -> Color? {
        guard let raw, !raw.isEmpty else { return nil }
        return hexColor(resolveThemed(raw, ctx: ctx))
    }

    private static func resolveThemed(_ raw: String, ctx: BindingContext) -> String {
        if ctx.env["theme"]?.stringValue == "dark", raw.hasPrefix("$token.color.") {
            let darkRef = raw.replacingOccurrences(of: "$token.color.", with: "$token.colorDark.")
            let darkValue = BindingEngine.resolveString(darkRef, in: ctx)
            if !darkValue.isEmpty { return darkValue }
        }
        return BindingEngine.resolveString(raw, in: ctx)
    }

    static func hexColor(_ hex: String) -> Color? {
        var s = hex.trimmingCharacters(in: .whitespaces)
        guard s.hasPrefix("#") else { return nil }
        s.removeFirst()
        guard let value = UInt64(s, radix: 16) else { return nil }
        let r, g, b, a: Double
        switch s.count {
        case 6:
            r = Double((value & 0xFF0000) >> 16) / 255
            g = Double((value & 0x00FF00) >> 8) / 255
            b = Double(value & 0x0000FF) / 255
            a = 1
        case 8:
            r = Double((value & 0xFF000000) >> 24) / 255
            g = Double((value & 0x00FF0000) >> 16) / 255
            b = Double((value & 0x0000FF00) >> 8) / 255
            a = Double(value & 0x000000FF) / 255
        default:
            return nil
        }
        return Color(.sRGB, red: r, green: g, blue: b, opacity: a)
    }

    // MARK: Typography

    /// Resolves a `$token.typography.*` ref into a SwiftUI `Font`.
    public static func font(_ styleRef: String?, ctx: BindingContext) -> Font? {
        guard let styleRef else { return nil }
        let node = BindingEngine.resolve(styleRef, in: ctx)
        guard let size = node["size"]?.doubleValue else { return nil }
        let weight = fontWeight(node["weight"]?.stringValue)
        return .system(size: size, weight: weight)
    }

    private static func fontWeight(_ name: String?) -> Font.Weight {
        switch name {
        case "bold": return .bold
        case "semibold": return .semibold
        case "medium": return .medium
        case "light": return .light
        default: return .regular
        }
    }
}
#endif

File: android/sdui/src/main/java/dev/sdui/render/Theme.kt
package dev.sdui.render

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.sdui.core.BindingContext
import dev.sdui.core.BindingEngine
import dev.sdui.core.Dimension

/**
 * Resolves contract values (`Dimension`, hex colours, typography tokens) into
 * native Compose types, using the shared token table carried by the binding
 * context.
 *
 * All platform-specific mapping lives here so the renderers stay declarative —
 * the Compose counterpart of the iOS `Theme` enum. Where SwiftUI returns
 * `CGFloat`/`Color`/`Font`, this returns `Dp`/`Color`/`TextStyle`.
 */
object Theme {

    /** The Compose sentinel meaning "fill the available space" for a [Dimension]. */
    val INFINITY: Dp = Dp.Infinity

    // MARK: Dimensions

    /**
     * Resolves a [Dimension] to a Compose [Dp]. Token refs are looked up in the
     * binding context; a missing token falls back to [default]. `infinity` maps
     * to [Dp.Infinity] so `frame.maxWidth = "infinity"` fills the axis.
     */
    fun dp(dim: Dimension?, ctx: BindingContext, default: Dp = 0.dp): Dp = when (dim) {
        null -> default
        is Dimension.Points -> dim.value.dp
        Dimension.Infinity -> Dp.Infinity
        is Dimension.Token ->
            (BindingEngine.resolve(dim.ref, ctx).doubleValue ?: default.value.toDouble()).dp
    }

    // MARK: Colors

    /**
     * Resolves a colour reference to a Compose [Color], or `null` when it cannot
     * be parsed. Accepts a token ref (`$token.color.primary`), a hex string
     * (`#RRGGBB` / `#RRGGBBAA`), or a binding — resolving through the token table
     * exactly like the iOS `Theme.color`.
     */
    fun color(raw: String?, ctx: BindingContext): Color? {
        if (raw.isNullOrEmpty()) return null
        val resolved = BindingEngine.resolveString(raw, ctx)
        return hexColor(resolved)
    }

    /**
     * The app's accent — the design token `color.primary`, resolved fresh so a live theme
     * swap re-tints everything. This is the single fallback for every default a component
     * needs where iOS draws with SwiftUI's `.accentColor`, replacing the scattered
     * hardcoded blues/purples that diverged from the theme (parity #35).
     */
    fun accent(ctx: BindingContext): Color = color("\$token.color.primary", ctx) ?: Color(0xFF5B5BF0)

    /** Parses `#RRGGBB` or `#RRGGBBAA` into a [Color], or `null` when malformed. */
    fun hexColor(hex: String): Color? {
        val s = hex.trim()
        if (!s.startsWith("#")) return null
        val body = s.substring(1)
        val value = body.toLongOrNull(16) ?: return null
        return when (body.length) {
            6 -> {
                val r = ((value and 0xFF0000) shr 16) / 255.0
                val g = ((value and 0x00FF00) shr 8) / 255.0
                val b = (value and 0x0000FF) / 255.0
                Color(r.toFloat(), g.toFloat(), b.toFloat(), 1f)
            }
            8 -> {
                val r = ((value and 0xFF000000) shr 24) / 255.0
                val g = ((value and 0x00FF0000) shr 16) / 255.0
                val b = ((value and 0x0000FF00) shr 8) / 255.0
                val a = (value and 0x000000FF) / 255.0
                Color(r.toFloat(), g.toFloat(), b.toFloat(), a.toFloat())
            }
            else -> null
        }
    }

    // MARK: Typography

    /**
     * Resolves a `$token.typography.*` ref into a Compose [TextStyle], or `null`
     * when the token has no `size`. Mirrors the iOS `Theme.font`, mapping the
     * shared `{size, weight}` descriptor to a system font.
     */
    fun textStyle(styleRef: String?, ctx: BindingContext): TextStyle? {
        styleRef ?: return null
        val node = BindingEngine.resolve(styleRef, ctx)
        val size = node["size"]?.doubleValue ?: return null
        return TextStyle(
            fontSize = size.sp,
            fontWeight = fontWeight(node["weight"]?.stringValue),
        )
    }

    private fun fontWeight(name: String?): FontWeight = when (name) {
        "bold" -> FontWeight.Bold
        "semibold" -> FontWeight.SemiBold
        "medium" -> FontWeight.Medium
        "light" -> FontWeight.Light
        else -> FontWeight.Normal
    }
}
